clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcher.java

```diff
@@ -278,55 +273,18 @@ private void validatePositionsAsync(Map<TopicPartition, FetchPosition> partition
             future.addListener(new RequestFutureListener<OffsetForEpochResult>() {
                 @Override
                 public void onSuccess(OffsetForEpochResult offsetsResult) {
-                    List<SubscriptionState.LogTruncation> truncations = new ArrayList<>();
-                    if (!offsetsResult.partitionsToRetry().isEmpty()) {
-                        subscriptions.setNextAllowedRetry(offsetsResult.partitionsToRetry(), time.milliseconds() + retryBackoffMs);
-                        metadata.requestUpdate();
-                    }
-
-                    // For each OffsetsForLeader response, check if the end-offset is lower than our current offset
-                    // for the partition. If so, it means we have experienced log truncation and need to reposition
-                    // that partition's offset.
-                    //
-                    // In addition, check whether the returned offset and epoch are valid. If not, then we should reset
-                    // its offset if reset policy is configured, or throw out of range exception.
-                    offsetsResult.endOffsets().forEach((topicPartition, respEndOffset) -> {
-                        FetchPosition requestPosition = fetchPositions.get(topicPartition);
-                        Optional<SubscriptionState.LogTruncation> truncationOpt =
-                                subscriptions.maybeCompleteValidation(topicPartition, requestPosition, respEndOffset);
-                        truncationOpt.ifPresent(truncations::add);
-                    });
-
-                    if (!truncations.isEmpty()) {
-                        offsetFetcherUtils.maybeSetOffsetForLeaderException(buildLogTruncationException(truncations));
-                    }
+                    offsetFetcherUtils.onSuccessfulRequestForValidatingPositions(fetchPositions,
```

Comment:
should we implement a offsetFetcherRequestFutureListener and we could do addListener(new offsetFetcherRequestListner())?

Reply:
This OffsetFetcher is the current code used by the KafkaConsumer, so I'm trying to minimize the changes, refactor it as needed for CTR by extracting code to OffsetFetcherUtils, but without changes in logic. That being said, we could definitely change the style here if we think it brings value, but keeping in mind the risk of changing what is currently working for the KafkaConsumer. WDYT?

Reply:
Sounds good thanks!